File: Hot100/src/com/action/double_pointer/ThreeSum.java
package com.action.double_pointer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {

    /*
    source: https://leetcode.cn/problems/3sum
    desc:
        给你一个整数数组 nums ，判断是否存在三元组 [nums[i], nums[j], nums[k]]
        满足 i != j、i != k 且 j != k ，
        同时还满足 nums[i] + nums[j] + nums[k] == 0 。
        请你返回所有和为 0 且不重复的三元组。
    * */


    /*
    首先将数组排序
    外层循环指向第一个数，使用双指针指向第二个数和第三个数
    若三数之和小于0，则移动左指针，使结果增大；若大于0，移动右指针，使结果减小
    */
    public static List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        // 固定最左边元素
        for (int i=0; i<nums.length-2; i++) {
            // 去除重复
            if (i>0 && nums[i]==nums[i-1]) continue;
            int left=i+1, right=nums.length-1;
            while(left < right){
                int sum = nums[i] + nums[left] + nums[right];
                if (sum < 0) {
                    left++;
                }else if (sum > 0) {
                    right--;
                }else {
                    ans.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    // 移动，避免重复
                    int value = nums[left];
                    while(left<right && nums[left]==value) left++;
                }
            }
        }
        return ans;
    }


    public static void main(String[] args) {
        int[] nums = {-1, 0, 1, 2, -1, -4};
        List<List<Integer>> ans = threeSum(nums);
        for (List<Integer> list : ans) {
            for (Integer integer : list) {
                System.out.print(integer + " ");
            }
            System.out.println();
        }
    }
}

File: doublepointer/three_sum.go
package doublepointer

import "sort"

/* 
15. 三数之和：https://leetcode.cn/problems/3sum/

给你一个整数数组 nums ，
判断是否存在三元组 [nums[i], nums[j], nums[k]] 满足 i != j、i != k 且 j != k ，
同时还满足 nums[i] + nums[j] + nums[k] == 0 。
请你返回所有和为 0 且不重复的三元组。

注意：答案中不可以包含重复的三元组。

示例 1：
	输入：nums = [-1,0,1,2,-1,-4]
	输出：[[-1,-1,2],[-1,0,1]]
	解释：
	nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0 。
	nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0 。
	nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0 。
	不同的三元组是 [-1,0,1] 和 [-1,-1,2] 。 
*/


// 先将数组进行排序，然后使用三个指针 i, left, right 分别指向三个数
// 保持 i 不变，若总和小于 0，移动 left，大于 0，移动 right
func ThreeSum(nums []int) [][]int {
    sort.Slice(nums, func(i, j int) bool {
        return nums[i] < nums[j]
    })
    ans := make([][]int, 0)
    for i:=0; i<len(nums); i++ {
        // 外层去重
        if i>0 && nums[i]==nums[i-1] {
            continue
        }
        left, right := i+1, len(nums)-1
        for (left < right) {
            sum := nums[i] + nums[left] + nums[right]
            if sum < 0 {    // 移动左指针，总和增大
                left++
            }else if sum > 0 {
                right--     // 移动右指针，总和减小
            }else {
                ans = append(ans, []int{nums[i], nums[left], nums[right]})
                base := nums[left]
                for left < right && nums[left]==base {
                    left++  // 内层去重
                }
            }
        }
    }
    return ans
}
